README.md
# clj-shopping

_Clojure is hard, let's go shopping_

This is a toy project, written while I was learning Clojure.

It is a small webservice that handles updates to a single shopping list.

### To retrieve the shopping list:

```
curl "http://localhost:3000/list"
```

### To modify the shopping list:

```
curl --request PATCH "http://localhost:3000/list?action=add&shop=Supermarkt&item=Milch"
```

## License

Copyright © 2015 Lars Immisch

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2015-11-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2015-11-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/clj-shopping/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clj-shopping/compare/0.1.0...0.1.1


test/clj_shopping/handler_test.clj
(ns clj-shopping.handler-test
  (:require [clojure.test :refer :all]
            [ring.mock.request :refer :all]
            [clj-shopping.handler :refer :all]
            [cheshire.core :as json]))

(defn json-body
  [response]
  (json/parse-string (:body response)))

(deftest test-operations
  (testing "add to list"
    (is (= (add-to-list {} "Supermarkt" "Senf") {"Supermarkt" ["Senf"]}))
    ;; Don't add duplicates
    (is (= (add-to-list {"Supermarkt" ["Senf"]} "Supermarkt" "Senf") {"Supermarkt" ["Senf"]}))
    (is (= (add-to-list {"Supermarkt" ["Senf"]} "Supermarkt" "Milch") {"Supermarkt" ["Senf" "Milch"]}))
    )

  (testing "done from list"
    (is (= (done-from-list {"Supermarkt" ["Milch" "Senf"]} "Supermarkt" "Senf") {"Supermarkt" ["Milch"]}))
    (is (= (done-from-list {"Supermarkt" ["Senf"]} "Supermarkt" "Senf") {}))
    )

  (testing "operations on the shopping-list ref"
    (dosync (ref-set shopping-list {}))
    (is (= (list-op add-to-list "Supermarkt" "Milch")) {"Supermarkt" ["Milch"]} )
    (is (= (list-op done-from-list "Supermarkt" "Milch")) {} )
    )
  )

(deftest test-app
  (testing "list patch endpoint"
    (dosync (ref-set shopping-list {}))
    (let [response (app (request :patch "/list"))]
      (is (= (:status response) 422))
      )
    (let [response (app (request :patch "/list" {"action" "add"}))]
      (is (= (:status response) 422))
      )
    (let [response (app (request :patch "/list" {"action" "add", "shop" "Supermarkt"}))]
      (is (= (:status response) 422))
      )
    (let [response (app (request :patch "/list" {"action" "add", "shop" "Supermarkt", "item", "Senf"}))]
      (is (= (:status response) 200))
      )
    )

  (testing "list get endpoint"
    (dosync (ref-set shopping-list {"Supermarkt" ["Senf"]}))
    (let [response (app (request :get "/list"))]
      (is (= (:status response) 200))
      (is (= (get-in response [:headers "Content-Type"]) "application/json"))
      (is (= (json-body response) {"Supermarkt" ["Senf"]}))
      )
    )

  (testing "list operations in sequence"
    (dosync (ref-set shopping-list {"Supermarkt" ["Senf"]}))
    (let [response (app (request :patch "/list" {"action" "add", "shop" "Supermarkt", "item", "Senf"}))]
      (is (= (:status response) 200)))
    (let [response (app (request :get "/list"))]
      (is (= (json-body response) {"Supermarkt" ["Senf"]}))
      )
    (let [response (app (request :patch "/list" {"action" "done", "shop" "Supermarkt", "item", "Senf"}))]
      (is (= (:status response) 200)))
    (let [response (app (request :get "/list"))]
      (is (= (json-body response) {}))
      )
    )

    (testing "not-found route"
    (let [response (app (request :get "/bogus-route"))]
      (is (= (:status response) 404))))
  )


src/clj_shopping/handler.clj
(ns clj-shopping.handler
  (:require [compojure.core :refer :all]
            [compojure.route :as route]
            [compojure.handler :as handler]
            [ring.middleware.json :refer :all]
            [ring.middleware.params :refer :all]
            [cheshire.core :as json]))

(def shopping-list (ref nil))

(defn init
  []
  (dosync
    (ref-set shopping-list (json/parse-stream (clojure.java.io/reader "shopping.json")))
  ))

(defn contains-value? [coll element]
  (boolean (some #(= element %) coll)))

(defn status-response
  [status body]
  {
   :status  status
   :headers {"Content-Type" "application/json"}
   :body    body
   })


(defn done-from-list
  [list shop item]
  (if-let [items (list shop)]
    (let [newitems (filter #(not= item %) items)]
      ;; remove empty shop mapping
      (if (empty? newitems)
        (dissoc list shop)
        (assoc list shop newitems)
        )
      )
    list
    )
  )

(defn add-to-list
  [list shop item]
  (let [items (or (list shop) [])]
    ;; don't add the same item twice
    ;; - if we used distinct, we'd convert the vector to a sequence
    (if (contains-value? items item)
      list
      (assoc list shop (conj items item)))
    )
  )

(defn list-op
  [op shop item]
  (dosync
    (alter shopping-list op shop item)
    )
  )

(defn list-patch-handler
  [{{action "action" shop "shop" item "item"} :params}]
  (cond
    (not (and shop item)) (status-response 422 {:message "Invalid/missing shop/item"})
    (= action "add") (status-response 200 ((partial list-op add-to-list) shop item))
    (= action "done") (status-response 200 ((partial list-op done-from-list) shop item))
    :else (status-response 422 {:message "Invalid/missing action"}))
  )

(defn list-handler
  [_]
  (status-response 200 @shopping-list)
  )

(defroutes app-routes
           (GET "/list" [] list-handler)
           (PATCH "/list" [] list-patch-handler)
           (route/not-found {:message "Page not found"}))

(defn wrap-log-request [handler]
  (fn [req]
    (println req)
    (handler req)))

(def app
  (-> app-routes
      wrap-log-request
      wrap-params
      wrap-json-response
      wrap-json-body))
